org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/util/JavaProjectClasspathChangeAnalyzer.java

```diff
@@ -84,6 +102,52 @@ public boolean isAttachmentChangeOnly(IJavaElementDelta delta) {
 		return true;
 	}
 	
+	private static Version installedJdtCoreVersion;
+	
+	private static boolean isJdtCoreGreaterOrEqual(Version version) {
+		if (installedJdtCoreVersion == null) {
+			installedJdtCoreVersion = JavaCore.getPlugin().getBundle().getVersion();
+		}
+		return installedJdtCoreVersion.compareTo(version) >= 0;
+	}
+
+	/**
+	 * this method checks if the change is a source or javadoc (de)attachment change only.
+	 * @since 2.30
+	 */
+	@Beta
+	protected boolean isAttachmentChangeFlagOnly(IJavaElementDelta child) {
+		int flags = child.getFlags();
+		if (flags == 0) {
+			return false;
+		}
+		int flagsWithoutSourceAttachments = flags & ~(IJavaElementDelta.F_SOURCEATTACHED | IJavaElementDelta.F_SOURCEDETACHED);
```

Comment:
```suggestion
		int F_CLASSPATH_ATTRIBUTES = 0x800000; // TODO use IJavaElementDelta.F_CLASSPATH_ATTRIBUTES
		int allOtherFlags = flags & ~(IJavaElementDelta.F_SOURCEATTACHED | IJavaElementDelta.F_SOURCEDETACHED | F_CLASSPATH_ATTRIBUTES); // TODO use IJavaElementDelta.F_CLASSPATH_ATTRIBUTES
		if (allOtherFlags != 0) {
			return false;
		}
		if (flags == F_CLASSPATH_ATTRIBUTES) {
```

Reply:
this would also not work as you can have source and javadoc attachment at the same time

Reply:
```
	protected boolean isAttachmentChangeFlagOnly(IJavaElementDelta child) {
		int flags = child.getFlags();
		if (flags == 0) {
			return false;
		}
		int F_CLASSPATH_ATTRIBUTES = 0x800000; // TODO use IJavaElementDelta.F_CLASSPATH_ATTRIBUTES
		int allOtherFlags = flags & ~(IJavaElementDelta.F_SOURCEATTACHED | IJavaElementDelta.F_SOURCEDETACHED | F_CLASSPATH_ATTRIBUTES); 
		if (allOtherFlags != 0) {
			// there are other flags present
			return false;
		}
		if ((flags & F_CLASSPATH_ATTRIBUTES) != 0) {
			// this is an attribute only change
			// TODO analyze child.getAttributeDeltas()
			try {
				// TODO dont use reflection when minimal target is 2023-03
				Object[] attributeDeltas = (Object[]) reflectExtensions.invoke(child, "getClasspathAttributeDeltas");
				for (Object attributeDelta : attributeDeltas) {
					String name = (String) reflectExtensions.invoke(attributeDelta, "getAttributeName");
					if (!IClasspathAttribute.JAVADOC_LOCATION_ATTRIBUTE_NAME.equals(name)) {
						return false;
					}
				}
			} catch (ReflectiveOperationException e) {
				logger.error("Something went wrong with the reflection code to read classpath attribute deltas", e);
				return false;
			}
			return true;
		}
```
???

Reply:
I don't think you need this here:
```
Object[] attributeDeltas = (Object[]) reflectExtensions.invoke(child, "getClasspathAttributeDeltas");
```
Since this check will not evaluate to `true` unless the API also exists?
```
if ((flags & F_CLASSPATH_ATTRIBUTES) != 0) {
```

Reply:
@trancexpress  i need to compile with old code, thus i need to read attributes and names with reflection

Reply:
> @trancexpress i need to compile with old code, thus i need to read attributes and names with reflection
Ah right, never mind then. I thought you were using reflection to check if the method exist.

Reply:
> ???

Yes, I think so. The last `return true` probably one line below.

Reply:
i wanted to keep two returns intentionally. should i add an else for the 2nd?